In [18]:
#!pip install pandas
#!pip install numpy

In [19]:
import pandas as pd
import os
import numpy as np

#### Merge the 12 months of sales date into a single CSV file

In [8]:
files = [file for file in os.listdir("/Users/trinhbich/Desktop/Python-portfolio-project/Sales_Data")]

all_months_data = pd.DataFrame()

# make a loop to concat the data
for file in files: 
    df = pd.read_csv("/Users/trinhbich/Desktop/Python-portfolio-project/Sales_Data/" + file) 
    all_months_data = pd.concat([all_months_data, df])

# export all data to csv
all_months_data.to_csv("all_data.csv", index=False)



#### Read in updated DataFrame

In [10]:
all_data = pd.read_csv("all_data.csv")
all_data.head()

Order ID               Product Quantity Ordered Price Each      Order Date  \
0   295665    Macbook Pro Laptop                1       1700  12/30/19 00:01   
1   295666    LG Washing Machine                1      600.0  12/29/19 07:03   
2   295667  USB-C Charging Cable                1      11.95  12/12/19 18:21   
3   295668      27in FHD Monitor                1     149.99  12/22/19 15:13   
4   295669  USB-C Charging Cable                1      11.95  12/18/19 12:38   

                         Purchase Address  
0  136 Church St, New York City, NY 10001  
1     562 2nd St, New York City, NY 10001  
2    277 Main St, New York City, NY 10001  
3     410 6th St, San Francisco, CA 94016  
4           43 Hill St, Atlanta, GA 30301

### Data Preprocessing

Data information

In [11]:
# Getting the information
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 8.6+ MB


Variables category

In [12]:
categorical = all_data.select_dtypes(['category','object']).columns
for col in categorical:
    print('{} : {} unique value(s)'.format(col,all_data[col].nunique()))

Order ID : 178438 unique value(s)
Product : 20 unique value(s)
Quantity Ordered : 10 unique value(s)
Price Each : 24 unique value(s)
Order Date : 142396 unique value(s)
Purchase Address : 140788 unique value(s)


Missing data points

In [15]:
# How many missing data points per columns?
missing_values_count = all_data.isnull().sum()

# look at the # of missing points in the first 10 columns
missing_values_count[0:10]

Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
dtype: int64

Whate percentage of the values in our dataset?

In [25]:
total_cells = np.product(all_data.shape)
total_missing = missing_values_count.sum()

percent_missing = (total_missing / total_cells)*100
print(f"{percent_missing: .2f}%")

 0.29%


### Clean up the data

#### Drop rows of NAN

In [27]:
# NAN values
nan_df = all_data[all_data.isna().any(axis=1)]
nan_df.head()
# Drop the rows of NAN data
all_data = all_data.dropna(how='all')
# Clean Future Warnings 'OR'
all_data = all_data[all_data['Order Date'].str[0:2] !='Or']
all_data

Order ID                 Product Quantity Ordered Price Each  \
0        295665      Macbook Pro Laptop                1       1700   
1        295666      LG Washing Machine                1      600.0   
2        295667    USB-C Charging Cable                1      11.95   
3        295668        27in FHD Monitor                1     149.99   
4        295669    USB-C Charging Cable                1      11.95   
...         ...                     ...              ...        ...   
186845   222905  AAA Batteries (4-pack)                1       2.99   
186846   222906        27in FHD Monitor                1     149.99   
186847   222907    USB-C Charging Cable                1      11.95   
186848   222908    USB-C Charging Cable                1      11.95   
186849   222909  AAA Batteries (4-pack)                1       2.99   

            Order Date                        Purchase Address  
0       12/30/19 00:01  136 Church St, New York City, NY 10001  
1       12/29/19 07:03     562 2nd St, New York City, NY 10001  
2       12/12/19 18:21    277 Main St, New York City, NY 10001  
3       12/22/19 15:13     410 6th St, San Francisco, CA 94016  
4       12/18/19 12:38           43 Hill St, Atlanta, GA 30301  
...                ...                                     ...  
186845  06/07/19 19:02           795 Pine St, Boston, MA 02215  
186846  06/01/19 19:29   495 North St, New York City, NY 10001  
186847  06/22/19 18:57   319 Ridge St, San Francisco, CA 94016  
186848  06/26/19 18:35    916 Main St, San Francisco, CA 94016  
186849  06/25/19 14:33          209 11th St, Atlanta, GA 30301  

[185950 rows x 6 columns]

### Convert columns to the correct type

In [7]:
all_data['Quantity Ordered'] = pd.to_numeric(all_data['Quantity Ordered'])# Make int
all_data['Price Each'] = pd.to_numeric(all_data['Price Each']) # Make float


all_data.head()

Order ID               Product  Quantity Ordered  Price Each  \
0   295665    Macbook Pro Laptop                 1     1700.00   
1   295666    LG Washing Machine                 1      600.00   
2   295667  USB-C Charging Cable                 1       11.95   
3   295668      27in FHD Monitor                 1      149.99   
4   295669  USB-C Charging Cable                 1       11.95   

       Order Date                        Purchase Address  
0  12/30/19 00:01  136 Church St, New York City, NY 10001  
1  12/29/19 07:03     562 2nd St, New York City, NY 10001  
2  12/12/19 18:21    277 Main St, New York City, NY 10001  
3  12/22/19 15:13     410 6th St, San Francisco, CA 94016  
4  12/18/19 12:38           43 Hill St, Atlanta, GA 30301

#### Convert Quantity Ordered column and Price Each column to categorical data type

In [29]:
# Convert the data
all_data['Quantity Ordered'], all_data['Price Each'] = all_data['Quantity Ordered'].astype('int64'), all_data['Price Each'].astype('float')

# Check the resutl
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 185950 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Order ID          185950 non-null  object 
 1   Product           185950 non-null  object 
 2   Quantity Ordered  185950 non-null  int64  
 3   Price Each        185950 non-null  float64
 4   Order Date        185950 non-null  object 
 5   Purchase Address  185950 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 9.9+ MB


#### Convert Order Date column to datetime data type

In [30]:
all_data['Order Date'] = pd.to_datetime(all_data['Order Date'])
all_data

Order ID                 Product  Quantity Ordered  Price Each  \
0        295665      Macbook Pro Laptop                 1     1700.00   
1        295666      LG Washing Machine                 1      600.00   
2        295667    USB-C Charging Cable                 1       11.95   
3        295668        27in FHD Monitor                 1      149.99   
4        295669    USB-C Charging Cable                 1       11.95   
...         ...                     ...               ...         ...   
186845   222905  AAA Batteries (4-pack)                 1        2.99   
186846   222906        27in FHD Monitor                 1      149.99   
186847   222907    USB-C Charging Cable                 1       11.95   
186848   222908    USB-C Charging Cable                 1       11.95   
186849   222909  AAA Batteries (4-pack)                 1        2.99   

                Order Date                        Purchase Address  
0      2019-12-30 00:01:00  136 Church St, New York City, NY 10001  
1      2019-12-29 07:03:00     562 2nd St, New York City, NY 10001  
2      2019-12-12 18:21:00    277 Main St, New York City, NY 10001  
3      2019-12-22 15:13:00     410 6th St, San Francisco, CA 94016  
4      2019-12-18 12:38:00           43 Hill St, Atlanta, GA 30301  
...                    ...                                     ...  
186845 2019-06-07 19:02:00           795 Pine St, Boston, MA 02215  
186846 2019-06-01 19:29:00   495 North St, New York City, NY 10001  
186847 2019-06-22 18:57:00   319 Ridge St, San Francisco, CA 94016  
186848 2019-06-26 18:35:00    916 Main St, San Francisco, CA 94016  
186849 2019-06-25 14:33:00          209 11th St, Atlanta, GA 30301  

[185950 rows x 6 columns]

Recap Data:
- DataFrame has total 186.850 records and 6 columns cateogircal type
- The number of missing value accounts for 0.29%
- Order ID: 178438 unique value(s)
- Product: 20 unique value(s)
- Price Each: 24 unique value(s)
- Order Date: 142396 unique value(s)
- Purchase Address: 140788 unique value(s)

### Data preparation

##### Augment data with additional colums

 Add Month, Hour, Minute, Sales, Cities Column

In [48]:
def augment_data(data):
    # Get the City in 'Purchase Address'
    def get_city(address):
        return address.split(',')[1]
    
    # Get the State in 'Purchase Address'
    def get_state(address):
        return address.split(',')[2].split(' ')[1]
    
    # Get month data
    all_data['Month'] =all_data['Order Date'].dt.month
    
    # Get Hour data
    data['Hour'] = data['Order Date'].dt.hour
    
    # Get Minute data
    data['Minute'] = data['Order Date'].dt.minute
    
    # Add Sales column 
    data['Sale'] = data['Quantity Ordered'] * data['Price Each']
    
    # Get Cities Data
    data['Cities'] = data['Purchase Address'].apply(lambda x: f"{get_city(x)} ({get_state(x)})")
    return data

    all_data = augment_data(all_data)
    all_data.head()
    

#### Task 2: Add Month Column

In [49]:
all_data['Month'] = all_data['Order Date'].str(tdate.time())[0:2]
all_data['Month'] = all_data['Month'].astype('int32')
all_data.head()

AttributeError: Can only use .str accessor with string values!

#### Task 3: Add a sales column

In [9]:
all_data['Sales'] = all_data["Quantity Ordered"] = all_data['Price Each']
all_data.head()

Order ID               Product  Quantity Ordered  Price Each  \
0   295665    Macbook Pro Laptop           1700.00     1700.00   
1   295666    LG Washing Machine            600.00      600.00   
2   295667  USB-C Charging Cable             11.95       11.95   
3   295668      27in FHD Monitor            149.99      149.99   
4   295669  USB-C Charging Cable             11.95       11.95   

       Order Date                        Purchase Address  Month    Sales  
0  12/30/19 00:01  136 Church St, New York City, NY 10001     12  1700.00  
1  12/29/19 07:03     562 2nd St, New York City, NY 10001     12   600.00  
2  12/12/19 18:21    277 Main St, New York City, NY 10001     12    11.95  
3  12/22/19 15:13     410 6th St, San Francisco, CA 94016     12   149.99  
4  12/18/19 12:38           43 Hill St, Atlanta, GA 30301     12    11.95

# Question 1: What was the best month for sales? How much was earned that month?

In [33]:
all_data.groupby('Month').sum()

KeyError: 'Month'

In [31]:
import matplotlib.pyplot as plt
months = range(1,13)
plt.bar(months, results['Sales'])
plt.xticks(months)
plt.show()


NameError: name 'results' is not defined